src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs

```diff
@@ -177,33 +178,51 @@ public virtual bool IsSupportedAlgorithm(string algorithm, SecurityKey key)
                 if (jsonWebKey.Kty == JsonWebAlgorithmsKeyTypes.RSA)
                     return IsRsaAlgorithmSupported(algorithm);
                 else if (jsonWebKey.Kty == JsonWebAlgorithmsKeyTypes.EllipticCurve)
-                    return IsEcdsaAlgorithmSupported(algorithm);
+                {
+#if NETSTANDARD1_4
+                    ECDsa ecdsa = AsymmetricSignatureProvider.CreateECDsaFromJsonWebKey(jsonWebKey, isPrivateKey);
+#else
+                    ECDsaCng ecdsa = AsymmetricSignatureProvider.CreateECDsaFromJsonWebKey(jsonWebKey, isPrivateKey);
+#endif
+                    return IsEcdsaAlgorithmSupported(algorithm, ecdsa);
+                }
                 else if (jsonWebKey.Kty == JsonWebAlgorithmsKeyTypes.Octet)
                     return IsSymmetricAlgorithmSupported(algorithm);
 
                 return false;
             }
 
-            if (key as ECDsaSecurityKey != null)
-                return IsEcdsaAlgorithmSupported(algorithm);
+            ECDsaSecurityKey ecdsaSecurityKey = key as ECDsaSecurityKey;
+            if (ecdsaSecurityKey != null)
+                return IsEcdsaAlgorithmSupported(algorithm, ecdsaSecurityKey.ECDsa);
 
             if (key as SymmetricSecurityKey != null)
                 return IsSymmetricAlgorithmSupported(algorithm);
 
             return false;
         }
 
-        private bool IsEcdsaAlgorithmSupported(string algorithm)
+        private bool IsEcdsaAlgorithmSupported(string algorithm, ECDsa key)
         {
             switch (algorithm)
             {
                 case SecurityAlgorithms.EcdsaSha256:
-                case SecurityAlgorithms.EcdsaSha384:
-                case SecurityAlgorithms.EcdsaSha512:
                 case SecurityAlgorithms.EcdsaSha256Signature:
+                    if (key.KeySize == 256)
```

Comment:
I am **definitely not a crypto expert**, but I doubt the key size alone is a very strong indicator for this kind of check, as nothing prevents me from using another curve than the ones expected here (respectively P-256, P-384 and P-521): if the key size is 256/384/521, it will return `true` even if the curve doesn't correspond.

Consider comparing the curve OID instead of the key size here.